File: LeetCode/Find if Array Can Be Sorted.swift
/*
  Find if Array Can Be Sorted
  URL: https://leetcode.com/problems/find-if-array-can-be-sorted
  Difficulty: Medium
  Topic : Array, Bit Manipulation, Sorting
*/

class Solution {
    func canSortArray(_ nums: [Int]) -> Bool {
        var n = nums
        var setBit: [Int:Int] = [:]
        nums.forEach { num in
            setBit[num] = num.nonzeroBitCount
        }
        
        for _ in 0..<nums.count-1 {
            var flag = true
            for j in 0..<nums.count-1 {
                if n[j] > n[j+1] {
                    if setBit[n[j]]! != setBit[n[j+1]]! {
                        return false
                    }
                    let temp = n[j]
                    n[j] = n[j+1]
                    n[j+1] = temp
                    flag = false
                }
            }
            if flag { break }
        }
        return true
    }
}

File: LeetCode/Python/Find if Array Can Be Sorted.py
"""
  Find if Array Can Be Sorted
  URL: https://leetcode.com/problems/find-if-array-can-be-sorted
  Difficulty: Medium
  Topic : Array, Bit Manipulation, Sorting
"""

class Solution:
    def canSortArray(self, nums: List[int]) -> bool:
        setBit = {}
        for num in nums:
            setBit[num] = num.bit_count()
        for _ in range(len(nums)-1):
            flag = True
            for i in range(len(nums)-1):
                if nums[i] > nums[i+1]:
                    if setBit[nums[i]] != setBit[nums[i+1]]:
                        return False
                    temp = nums[i]
                    nums[i] = nums[i+1]
                    nums[i+1] = temp
                    flag = False
            if flag: break
        return True
